server/jetstream_benchmark_test.go

```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+		quitCh chan bool
+	}
+
+	messageSizeCases := []int64{
+		100,        // 100B
+		1024,       // 1KiB
+		10240,      // 10KiB
+		512 * 1024, // 512KiB
+	}
+	numPubsCases := []int{
+		3,
+		5,
+		10,
+	}
+
+	fanIn := func(b *testing.B, numPubs int, messageSize int64) {
+
+		cl, ls, shutdown, nc, js := startJSClusterAndConnect(b, clusterSize)
+		defer shutdown()
+		defer nc.Close()
+
+		clientUrl := ls.ClientURL()
+
+		_, err := js.AddStream(&nats.StreamConfig{
+			Name:     streamName,
+			Subjects: []string{subject},
+			Replicas: replicas,
+		})
+		if err != nil {
+			b.Fatal(err)
+		}
+		defer js.DeleteStream(streamName)
+
+		// If replicated resource, connect to stream leader for lower variability
+		if replicas > 1 {
+			nc.Close()
+			clientUrl = cl.streamLeader("$G", streamName).ClientURL()
+			nc, _ = jsClientConnectURL(b, clientUrl)
+			defer nc.Close()
+		}
+
+		// random bytes as payload
+		messageData := make([]byte, messageSize)
+		if _, err := rand.Read(messageData); err != nil {
+			b.Fatal(err)
+		}
+
+		// waits for all publishers sub-routines and for main thread to be ready
+		var publishersReadyWg sync.WaitGroup
+		publishersReadyWg.Add(numPubs + 1)
+
+		// create N publishers
+		publishers := func() []BenchPublisher {
+			pubs := make([]BenchPublisher, numPubs)
+			for i := range pubs {
+				// create a quit channel
+				pubs[i].quitCh = make(chan bool, 1)
+				// create publisher connection and jetstream context
+				ncPub, err := nats.Connect(clientUrl)
+				if err != nil {
+					b.Fatal(err)
+				}
+				pubs[i].conn = ncPub
+				jsPub, err := ncPub.JetStream()
+				if err != nil {
+					b.Fatal(err)
+				}
+				pubs[i].js = jsPub
+
+				// publisher successfully initialized
+				publishersReadyWg.Done()
+			}
+			return pubs
+		}()
+
+		// wait group to ensure all publishers have been torn down
+		var finishedPublishersWg sync.WaitGroup
+		finishedPublishersWg.Add(numPubs)
+
+		// wait group blocks main thread until publish workload is completed, it is decremented after stream receives b.N messages from all publishers
+		var benchCompleteWg sync.WaitGroup
+		benchCompleteWg.Add(1)
+
+		// start go routines for all publishers, wait till all publishers are initialized before starting publish workload
+		for i := range publishers {
+			go func(pubId int) {
+				// signal that this publisher has been torn down
+				defer finishedPublishersWg.Done()
+
+				// wait till all other publishers are ready to start workload
+				publishersReadyWg.Wait()
+
+				// publish until quitCh is closed
+				for {
+					select {
+					case <-publishers[pubId].quitCh:
+						publishers[pubId].conn.Close()
+						return
+					default:
+						// continue publishing
+					}
+					pubAck, err := publishers[pubId].js.Publish(subject, messageData)
+					if err != nil {
+						publishers[pubId].publishErrors += 1
+						fmt.Println(err)
+						continue
+					}
+
+					// all messages have been published to stream
+					if int(pubAck.Sequence) == b.N {
```

Comment:
Doesn't matter but as general rule it's probably safer to cast the smallest type to the larger, so cast `b.N`

Reply:
If you change this check to: `if pubAck.Sequence > b.N  then return` you can simplify your code quite a bit, no need for a quit channel.

Slightly changes what you are measuring, i.e. "time for everyone to see a puback > b.N".

You could bring it back to "time for stream to reach b.N" with some additional mechanism triggered by just 1 (the first) producer.